src/http/router/fs_dir.zig

```diff
@@ -0,0 +1,258 @@
+                std.posix.close(provision.fd);
+                try provision.context.close();
+                return;
+            }
+
+            const length: usize = @intCast(result);
+            provision.rd_offset += length;
+            provision.current_length += length;
+            log.debug("current offset: {d} | fd: {}", .{ provision.rd_offset, provision.fd });
+
+            if (provision.rd_offset >= provision.file_size or result == 0) {
+                log.debug("done streaming file | rd off: {d} | f size: {d} | result: {d}", .{
+                    provision.rd_offset,
+                    provision.file_size,
+                    result,
+                });
+
+                std.posix.close(provision.fd);
+                try provision.context.send_then_recv(provision.buffer[0..provision.current_length]);
+            } else {
+                assert(provision.current_length <= provision.buffer.len);
+                if (provision.current_length == provision.buffer.len) {
+                    try provision.context.send_then(
+                        provision.buffer[0..provision.current_length],
+                        provision,
+                        send_file_task,
+                    );
+                } else {
+                    try rt.fs.read(
+                        provision,
+                        read_file_task,
+                        provision.fd,
+                        provision.buffer[provision.current_length..],
+                        provision.rd_offset,
+                    );
+                }
+            }
+        }
+
+        fn send_file_task(rt: *Runtime, success: bool, provision: *FileProvision) !void {
+            errdefer {
+                std.posix.close(provision.fd);
+                provision.context.close() catch unreachable;
+            }
+
+            if (!success) {
+                log.warn("send file stream failed!", .{});
+                std.posix.close(provision.fd);
+                return;
+            }
+
+            // reset current length
+            provision.current_length = 0;
+
+            // continue streaming..
+            try rt.fs.read(
+                provision,
+                read_file_task,
+                provision.fd,
+                provision.buffer,
+                provision.rd_offset,
+            );
+        }
+
+        pub fn handler_fn(ctx: *Context, dir_path: []const u8) !void {
+            if (ctx.captures.len == 0) {
+                try ctx.respond(.{
+                    .status = .@"Not Found",
+                    .mime = Mime.HTML,
+                    .body = "",
+                });
+                return;
+            }
+
+            //TODO Can we do this once and for all at initialization?
+            // Resolving the base directory.
+            const resolved_dir = try std.fs.path.resolve(ctx.allocator, &[_][]const u8{ dir_path });
+            defer ctx.allocator.free(resolved_dir);
```

Comment:
This was previously made once and for all on route definition. However, now that route definition is entirely comptime, it's not possible. I could make something like a "route context" that would be initialized at runtime when running `router.init`, but it's not a small change so I wanted to know your opinion on that problem before doing it.

Reply:
For this PR, I think it's fine to leave it as (aka doing it twice) **BUT** you need to use realpath commands over resolve.

Resolve keeps the path as relative to the current position so you can do some weird stuff with `../` and `./`. The realpath one will convert it to the absolute version which will ensure the later path check is correct.

Reply:
I chose `resolve` over `realpath` because of this: https://github.com/ziglang/zig/issues/19353
(specified in the commit for reference, but I wasn't sure if I should make a comment or not)

> The realpath one will convert it to the absolute version which will ensure the later path check is correct.

I don't see why `resolve` is not doing that. Let's say that `dir_path` is "../something/./../webroot/.": resolve should return "../webroot".
Then, we do the same with `dir_path` + `search_path`: "../something/./../webroot/." then "../danger/secret.txt". Resolve should return "../danger/secret.txt". The check enforces that the resolved file path starts with the **exact same beginning**, which means no one could possibly go up in the path.

As far as I know, `resolve` and `realpath` are doing kind of the same thing, with the only difference that `resolve` will not try to add the current position to get an absolute path (we could reproduce `realpath` with `resolve` by just adding `cwd` as the first parameter, but it would be a no-op for the difference comparison that we do right after that).

I wrote a test to check my assumptions:

```zig
test "path checking" {
    const dir_path = "../something/./../webroot/.";
    const resolved_dir_path = try std.fs.path.resolve(std.testing.allocator, &[_][]const u8{dir_path});
    defer std.testing.allocator.free(resolved_dir_path);
    const search_path = "../danger/secret.txt";
    const resolved_file_path = try std.fs.path.resolve(std.testing.allocator, &[_][]const u8{dir_path, search_path});
    defer std.testing.allocator.free(resolved_file_path);

    // Resolve simplifies the path.
    try std.testing.expectEqualStrings("../webroot", resolved_dir_path);
    try std.testing.expectEqualStrings("../danger/secret.txt", resolved_file_path);

    // Not the same beginning.
    try std.testing.expect(!std.mem.startsWith(u8, resolved_file_path, resolved_dir_path));
}
```

I'm not against using `realpath` and I'll do it if you want it, but as `resolve` will be a bit more efficient (probably not significant if the `cwd` path is not large) and `realpath` risks to be removed in the future, using `resolve` seems a better choice.

Reply:
You are correct in this case and using resolve is the better option, my mistake. Thanks for double checking!

Reply:
OK so I'm keeping it as is and I'll try to do something about the "route resolution on every call" issue later. Thanks! :-)